xds/internal/xdsclient/xdsresource/unmarshal_lds.go

```diff
@@ -235,14 +236,13 @@ func processHTTPFilters(filters []*v3httppb.HttpFilter, server bool) ([]HTTPFilt
 	if len(ret) == 0 {
 		return nil, fmt.Errorf("http filters list is empty")
 	}
-	var i int
-	for ; i < len(ret)-1; i++ {
+	for i := 0; i < len(ret)-1; i++ {
 		if ret[i].Filter.IsTerminal() {
 			return nil, fmt.Errorf("http filter %q is a terminal filter but it is not last in the filter chain", ret[i].Name)
 		}
 	}
-	if !ret[i].Filter.IsTerminal() {
-		return nil, fmt.Errorf("http filter %q is not a terminal filter", ret[len(ret)-1].Name)
+	if !router.IsRouterFilter(ret[len(ret)-1].Filter) {
+		return nil, fmt.Errorf("http filter %q is not a router filter", ret[len(ret)-1].Name)
```

Comment:
Nit: Maybe we should indicate the fact that this is the last http filter. So, something like
`return nil, fmt.Errorf("last http filter in the list, %q, is not a router filter", ret[len(ret)-1].Name)`

Reply:
have replaced the error message with `fmt.Errorf("last http filter (%q) in the filter-chain is not a router filter", ret[len(ret)-1].Name)` locally
will push here with the next commit